### ISSUE TITLE
Shorter errors for lax primitives

### ISSUE THREAD

**Issue by joeryjoery** [google/jax#19527]
When I run snippets like `jax.lax.select` on incorrect shapes or dtypes, which should be an easy and fast check, I get absurdly long error messages.

Could this be improved? 

EDIT: jax version: 0.20.0 on Ubuntu WSL Python 3.10

Example,
```python
import jax
import jax.numpy as jnp


jax.lax.select(
    True,
    jnp.ones(3, jnp.int32),
    jnp.zeros(5, jnp.float32)
)

>>TypeError                                 Traceback (most recent call last)
Cell In[8], line 1
----> 1 jax.lax.select(
      2     True,
      3     jnp.ones(3, jnp.int32),
      4     jnp.zeros(5, jnp.float32)
      5 )

File ~/anaconda3/envs/jax3.10/lib/python3.10/site-packages/jax/_src/lax/lax.py:920, in select(pred, on_true, on_false)
    898 """Selects between two branches based on a boolean predicate.
    899 
    900 Wraps XLA's `Select
   (...)
    916   result: array with same shape and dtype as ``on_true`` and ``on_false``.
    917 """
    918 # Caution! The select_n_p primitive has the *opposite* order of arguments to
    919 # select(). This is because it implements `select_n`.
--> 920 return select_n_p.bind(pred, on_false, on_true)

File ~/anaconda3/envs/jax3.10/lib/python3.10/site-packages/jax/_src/core.py:385, in Primitive.bind(self, *args, **params)
    382 def bind(self, *args, **params):
    383   assert (not config.enable_checks.value or
    384           all(isinstance(arg, Tracer) or valid_jaxtype(arg) for arg in args)), args
--> 385   return self.bind_with_trace(find_top_trace(args), args, params)

File ~/anaconda3/envs/jax3.10/lib/python3.10/site-packages/jax/_src/core.py:388, in Primitive.bind_with_trace(self, trace, args, params)
    387 def bind_with_trace(self, trace, args, params):
--> 388   out = trace.process_primitive(self, map(trace.full_raise, args), params)
    389   return map(full_lower, out) if self.multiple_results else full_lower(out)

File ~/anaconda3/envs/jax3.10/lib/python3.10/site-packages/jax/_src/core.py:868, in EvalTrace.process_primitive(self, primitive, tracers, params)
    867 def process_primitive(self, primitive, tracers, params):
--> 868   return primitive.impl(*tracers, **params)

File ~/anaconda3/envs/jax3.10/lib/python3.10/site-packages/jax/_src/dispatch.py:128, in apply_primitive(prim, *args, **params)
    126   in_avals, in_shardings = util.unzip2([_arg_spec(a) for a in args])
    127   in_tree = tree_util.tree_structure(args)
--> 128   compiled_fun = xla_primitive_callable(
    129       prim, in_avals, in_tree, OrigShardings(in_shardings), **params)
    130 except pxla.DeviceAssignmentMismatchError as e:
    131   fails, = e.args

File ~/anaconda3/envs/jax3.10/lib/python3.10/site-packages/jax/_src/util.py:284, in cache.<locals>.wrap.<locals>.wrapper(*args, **kwargs)
    282   return f(*args, **kwargs)
    283 else:
--> 284   return cached(config.config._trace_context(), *args, **kwargs)

File ~/anaconda3/envs/jax3.10/lib/python3.10/site-packages/jax/_src/util.py:277, in cache.<locals>.wrap.<locals>.cached(_, *args, **kwargs)
    275 @functools.lru_cache(max_size)
    276 def cached(_, *args, **kwargs):
--> 277   return f(*args, **kwargs)

File ~/anaconda3/envs/jax3.10/lib/python3.10/site-packages/jax/_src/dispatch.py:157, in xla_primitive_callable(prim, in_avals, in_tree, orig_in_shardings, **params)
    155 wrapped_fun = lu.wrap_init(prim_fun)
    156 flat_fun, out_tree = api_util.flatten_fun_nokwargs(wrapped_fun, in_tree)
--> 157 computation = sharded_lowering(
    158     flat_fun, prim.name, donated_invars, keep_unused=False,
    159     inline=True, in_avals=in_avals, in_shardings=orig_in_shardings.shardings,
    160     lowering_parameters=mlir.LoweringParameters())
    161 compiled = computation.compile()
    162 if xla_extension_version >= 192:

File ~/anaconda3/envs/jax3.10/lib/python3.10/site-packages/jax/_src/dispatch.py:188, in sharded_lowering(fun, name, donated_invars, keep_unused, inline, in_avals, in_shardings, lowering_parameters)
    183 in_shardings_unspec = [UNSPECIFIED if i is None else i for i in in_shardings]
    185 # Pass in a singleton `UNSPECIFIED` for out_shardings because we don't know
    186 # the number of output avals at this stage. lower_sharding_computation will
    187 # apply it to all out_avals.
--> 188 return pxla.lower_sharding_computation(
    189     fun, 'jit', name, in_shardings_unspec, UNSPECIFIED, donated_invars,
    190     in_avals, keep_unused=keep_unused, inline=inline,
    191     devices_from_context=None,
    192     lowering_parameters=lowering_parameters)

File ~/anaconda3/envs/jax3.10/lib/python3.10/site-packages/jax/_src/profiler.py:340, in annotate_function.<locals>.wrapper(*args, **kwargs)
    337 @wraps(func)
    338 def wrapper(*args, **kwargs):
    339   with TraceAnnotation(name, **decorator_kwargs):
--> 340     return func(*args, **kwargs)
    341   return wrapper

File ~/anaconda3/envs/jax3.10/lib/python3.10/site-packages/jax/_src/interpreters/pxla.py:1971, in lower_sharding_computation(fun_or_jaxpr, api_name, fun_name, in_shardings, out_shardings, donated_invars, global_in_avals, keep_unused, inline, devices_from_context, lowering_parameters)
   1965 # 1. Trace to jaxpr and preprocess/verify it
   1966 auto_spmd_lowering = (
   1967     check_if_any_auto(in_shardings) if is_unspecified(out_shardings) else
   1968     check_if_any_auto(it.chain.from_iterable([in_shardings, out_shardings])))  # type: ignore
   1970 (closed_jaxpr, global_in_avals, global_out_avals, donated_invars,
-> 1971  kept_var_idx, name_stack) = _trace_to_jaxpr_and_dce(
   1972     fun_or_jaxpr, global_in_avals, api_name, fun_name, keep_unused,
   1973     donated_invars, auto_spmd_lowering)
   1974 jaxpr = closed_jaxpr.jaxpr
   1975 in_shardings = tuple(s for i, s in enumerate(in_shardings) if i in kept_var_idx)

File ~/anaconda3/envs/jax3.10/lib/python3.10/site-packages/jax/_src/interpreters/pxla.py:1672, in cache_wrap.<locals>.wrapped(f, *args, **kwargs)
   1670 def wrapped(f, *args, **kwargs):
   1671   if isinstance(f, lu.WrappedFun):
-> 1672     return _wrapped_with_lu_cache(f, *args, **kwargs)
   1673   else:
   1674     return _wrapped_with_weakref_lru_cache(f, *args, **kwargs)

File ~/anaconda3/envs/jax3.10/lib/python3.10/site-packages/jax/_src/linear_util.py:349, in cache.<locals>.memoized_fun(fun, *args)
    347   fun.populate_stores(stores)
    348 else:
--> 349   ans = call(fun, *args)
    350   cache[key] = (ans, fun.stores)
    352 return ans

File ~/anaconda3/envs/jax3.10/lib/python3.10/site-packages/jax/_src/interpreters/pxla.py:1697, in _trace_to_jaxpr_and_dce(fun_or_jaxpr, global_in_avals, api_name, fun_name, keep_unused, donated_invars, auto_spmd_lowering)
   1693 if isinstance(fun_or_jaxpr, lu.WrappedFun):
   1694   with dispatch.log_elapsed_time(
   1695       "Finished tracing + transforming {fun_name} in {elapsed_time} sec",
   1696       fun_name=str(name_stack), event=dispatch.JAXPR_TRACE_EVENT):
-> 1697     jaxpr, global_out_avals, consts = pe.trace_to_jaxpr_final(
   1698         fun_or_jaxpr, global_in_avals)
   1699 else:
   1700   assert isinstance(fun_or_jaxpr, core.ClosedJaxpr)

File ~/anaconda3/envs/jax3.10/lib/python3.10/site-packages/jax/_src/profiler.py:340, in annotate_function.<locals>.wrapper(*args, **kwargs)
    337 @wraps(func)
    338 def wrapper(*args, **kwargs):
    339   with TraceAnnotation(name, **decorator_kwargs):
--> 340     return func(*args, **kwargs)
    341   return wrapper

File ~/anaconda3/envs/jax3.10/lib/python3.10/site-packages/jax/_src/interpreters/partial_eval.py:2356, in trace_to_jaxpr_final(fun, in_avals, debug_info, keep_inputs)
   2354   main.jaxpr_stack = ()  # type: ignore
   2355   with core.new_sublevel():
-> 2356     jaxpr, out_avals, consts = trace_to_subjaxpr_dynamic(
   2357       fun, main, in_avals, keep_inputs=keep_inputs, debug_info=debug_info)
   2358   del fun, main
   2359 return jaxpr, out_avals, consts

File ~/anaconda3/envs/jax3.10/lib/python3.10/site-packages/jax/_src/interpreters/partial_eval.py:2300, in trace_to_subjaxpr_dynamic(fun, main, in_avals, keep_inputs, debug_info)
   2298 in_tracers = _input_type_to_tracers(trace.new_arg, in_avals)
   2299 in_tracers_ = [t for t, keep in zip(in_tracers, keep_inputs) if keep]
-> 2300 ans = fun.call_wrapped(*in_tracers_)
   2301 out_tracers = map(trace.full_raise, ans)
   2302 jaxpr, consts = frame.to_jaxpr(out_tracers)

File ~/anaconda3/envs/jax3.10/lib/python3.10/site-packages/jax/_src/linear_util.py:191, in WrappedFun.call_wrapped(self, *args, **kwargs)
    188 gen = gen_static_args = out_store = None
    190 try:
--> 191   ans = self.f(*args, **dict(self.params, **kwargs))
    192 except:
    193   # Some transformations yield from inside context managers, so we have to
    194   # interrupt them before reraising the exception. Otherwise they will only
    195   # get garbage-collected at some later time, running their cleanup tasks
    196   # only after this exception is handled, which can corrupt the global
    197   # state.
    198   while stack:

File ~/anaconda3/envs/jax3.10/lib/python3.10/site-packages/jax/_src/dispatch.py:149, in xla_primitive_callable.<locals>.prim_fun(*args)
    148 def prim_fun(*args):
--> 149   out = prim.bind(*args, **params)
    150   if prim.multiple_results:
    151     return out

File ~/anaconda3/envs/jax3.10/lib/python3.10/site-packages/jax/_src/core.py:385, in Primitive.bind(self, *args, **params)
    382 def bind(self, *args, **params):
    383   assert (not config.enable_checks.value or
    384           all(isinstance(arg, Tracer) or valid_jaxtype(arg) for arg in args)), args
--> 385   return self.bind_with_trace(find_top_trace(args), args, params)

File ~/anaconda3/envs/jax3.10/lib/python3.10/site-packages/jax/_src/core.py:388, in Primitive.bind_with_trace(self, trace, args, params)
    387 def bind_with_trace(self, trace, args, params):
--> 388   out = trace.process_primitive(self, map(trace.full_raise, args), params)
    389   return map(full_lower, out) if self.multiple_results else full_lower(out)

File ~/anaconda3/envs/jax3.10/lib/python3.10/site-packages/jax/_src/interpreters/partial_eval.py:1972, in DynamicJaxprTrace.process_primitive(self, primitive, tracers, params)
   1970 if primitive in custom_staging_rules:
   1971   return custom_staging_rules[primitive](self, *tracers, **params)
-> 1972 return self.default_process_primitive(primitive, tracers, params)

File ~/anaconda3/envs/jax3.10/lib/python3.10/site-packages/jax/_src/interpreters/partial_eval.py:1976, in DynamicJaxprTrace.default_process_primitive(self, primitive, tracers, params)
   1974 def default_process_primitive(self, primitive, tracers, params):
   1975   avals = [t.aval for t in tracers]
-> 1976   out_avals, effects = primitive.abstract_eval(*avals, **params)
   1977   # == serve as a "not xor" here.
   1978   if not (isinstance(out_avals, (tuple,list)) == primitive.multiple_results):

File ~/anaconda3/envs/jax3.10/lib/python3.10/site-packages/jax/_src/core.py:421, in _effect_free_abstract_eval.<locals>.abstract_eval_(*args, **kwargs)
    420 def abstract_eval_(*args, **kwargs):
--> 421   return abstract_eval(*args, **kwargs), no_effects

File ~/anaconda3/envs/jax3.10/lib/python3.10/site-packages/jax/_src/lax/utils.py:60, in standard_abstract_eval(prim, shape_rule, dtype_rule, weak_type_rule, named_shape_rule, *avals, **kwargs)
     58   return core.ConcreteArray(out.dtype, out, weak_type=weak_type)
     59 elif least_specialized is core.ShapedArray:
---> 60   return core.ShapedArray(shape_rule(*avals, **kwargs),
     61                           dtype_rule(*avals, **kwargs), weak_type=weak_type,
     62                           named_shape=named_shape_rule(*avals, **kwargs))
     63 elif least_specialized is core.DShapedArray:
     64   shape = shape_rule(*avals, **kwargs)

File ~/anaconda3/envs/jax3.10/lib/python3.10/site-packages/jax/_src/lax/lax.py:3515, in _select_shape_rule(which, *cases)
   3513 if any(case.shape != cases[0].shape for case in cases[1:]):
   3514   msg = "select cases must have the same shapes, got [{}]."
-> 3515   raise TypeError(msg.format(", ".join([str(c.shape) for c in cases])))
   3516 if which.shape and which.shape != cases[0].shape:
   3517   msg = ("select `which` must be scalar or have the same shape as cases, "
   3518          "got `which` shape {} but case shape {}.")

TypeError: select cases must have the same shapes, got [(5,), (3,)].
```

**Comment by jakevdp** [comment]
Thanks for the suggestion! Maybe we could add some of these abstract evaluation frames to the pieces hidden by `jax_traceback_filtering`.

**Comment by jakevdp** [comment]
Actually, it looks like this is already covered. With `JAX_TRACEBACK_FILTERING='auto'` (the default value) I get this when running this script:
```pytb
$ python tmp.py 
jax.errors.SimplifiedTraceback: For simplicity, JAX has removed its internal frames from the traceback of the following exception. Set JAX_TRACEBACK_FILTERING=off to include these.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/vanderplas/github/google/jax/tmp.py", line 5, in <module>
    jax.lax.select(
  File "/Users/vanderplas/github/google/jax/jax/_src/lax/lax.py", line 931, in select
    return select_n_p.bind(pred, on_false, on_true)
  File "/Users/vanderplas/github/google/jax/jax/_src/core.py", line 444, in bind
    return self.bind_with_trace(find_top_trace(args), args, params)
  File "/Users/vanderplas/github/google/jax/jax/_src/core.py", line 447, in bind_with_trace
    out = trace.process_primitive(self, map(trace.full_raise, args), params)
  File "/Users/vanderplas/github/google/jax/jax/_src/core.py", line 935, in process_primitive
    return primitive.impl(*tracers, **params)
  File "/Users/vanderplas/github/google/jax/jax/_src/dispatch.py", line 87, in apply_primitive
    outs = fun(*args)
TypeError: select cases must have the same shapes, got [(5,), (3,)].
```
Can you check whether you're setting `JAX_TRACEBACK_FILTERING` to something different? It looks like you're running this in a notebook, so it's possible that you're running under a version of IPython that does not support filtering the traceback; see the logic here: https://github.com/google/jax/blob/f34bcc326ba882a500c9348d7992ae4f580bc4c3/jax/_src/traceback_util.py#L138-L150

**Comment by dynamicwebpaige** [comment]
I experienced a similar traceback when using Colab (see below). @jakevdp, would it be possible for us to change that default setting for Colab?

```python
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
[<ipython-input-1-39b3b1990ec7>](https://localhost:8080/#) in <cell line: 5>()
      3 
      4 
----> 5 jax.lax.select(
      6     True,
      7     jnp.ones(3, jnp.int32),

5 frames
[/usr/local/lib/python3.10/dist-packages/jax/_src/lax/lax.py](https://localhost:8080/#) in select(pred, on_true, on_false)
    926   # Caution! The select_n_p primitive has the *opposite* order of arguments to
    927   # select(). This is because it implements `select_n`.
--> 928   return select_n_p.bind(pred, on_false, on_true)
    929 
    930 def select_n(which: ArrayLike, *cases: ArrayLike) -> Array:

[/usr/local/lib/python3.10/dist-packages/jax/_src/core.py](https://localhost:8080/#) in bind(self, *args, **params)
    442     assert (not config.enable_checks.value or
    443             all(isinstance(arg, Tracer) or valid_jaxtype(arg) for arg in args)), args
--> 444     return self.bind_with_trace(find_top_trace(args), args, params)
    445 
    446   def bind_with_trace(self, trace, args, params):

[/usr/local/lib/python3.10/dist-packages/jax/_src/core.py](https://localhost:8080/#) in bind_with_trace(self, trace, args, params)
    445 
    446   def bind_with_trace(self, trace, args, params):
--> 447     out = trace.process_primitive(self, map(trace.full_raise, args), params)
    448     return map(full_lower, out) if self.multiple_results else full_lower(out)
    449 

[/usr/local/lib/python3.10/dist-packages/jax/_src/core.py](https://localhost:8080/#) in process_primitive(self, primitive, tracers, params)
    933 
    934   def process_primitive(self, primitive, tracers, params):
--> 935     return primitive.impl(*tracers, **params)
    936 
    937   def process_call(self, primitive, f, tracers, params):

[/usr/local/lib/python3.10/dist-packages/jax/_src/dispatch.py](https://localhost:8080/#) in apply_primitive(prim, *args, **params)
     85     prev = lib.jax_jit.swap_thread_local_state_disable_jit(False)
     86     try:
---> 87       outs = fun(*args)
     88     finally:
     89       lib.jax_jit.swap_thread_local_state_disable_jit(prev)

    [... skipping hidden 19 frame]

[/usr/local/lib/python3.10/dist-packages/jax/_src/lax/lax.py](https://localhost:8080/#) in _select_shape_rule(which, *cases)
   3520   if any(case.shape != cases[0].shape for case in cases[1:]):
   3521     msg = "select cases must have the same shapes, got [{}]."
-> 3522     raise TypeError(msg.format(", ".join([str(c.shape) for c in cases])))
   3523   if which.shape and which.shape != cases[0].shape:
   3524     msg = ("select `which` must be scalar or have the same shape as cases, "

TypeError: select cases must have the same shapes, got [(5,), (3,)].
```

**Comment by jakevdp** [comment]
Thanks Paige – it looks like things are working as expected there (note the`skipping hidden 19 frame` in your output). The traceback hiding code uses different mechanisms depending on whether it's being run in IPython or run in a script.

**Comment by jakevdp** [comment]
@joeryjoery it looks like you may be using an older JAX version as well. These tracebacks have been improved in recent releases, you might try updating JAX to see if the error is more useful. Thanks!

**Comment by joeryjoery** [comment]
Hey thanks for the quick reply! Yes forgot to state that my jax version is `0.20.0` I edited it in the post; I tested this in a IPython shell.

I tried setting `os.environ['JAX_TRACEBACK_FILTERING']` to all shown options: `["off", "tracebackhide", "remove_frames", "quiet_remove_frames", "auto"]`, but none of the options did anything. 


At the moment v:0.20.0 is stable for my PC, on newer versions my GPU doesn't get detected by an unrelated error: cuBLAS version mismatch.

But you're correct, when I tested this on another system where I do have jax v 0.23.0, the error message is automatically shortened. So I guess this is already fixed :). Thanks!
